### PR TITLE
[graph_trainer] Add make_fx_tracer utility and unit tests

### PR DESCRIPTION
**Pull request by ydwu4**
- nonstrict_tracer.py: make_fx-based tracer with subclass handling (trace_module, run_traced_module, SubclassMeta, unwrap/wrap helpers)
- test_trace_module.py: simple MLP tests verifying bitwise-exact eager vs traced execution over multiple training steps

cc @SherlockNoMad

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
let's not call this nonstrict tracer, this is a scary name. I would prefer the modest "make_fx_tracer" or "minimal_tracer" (but this is reserved for MVT?)

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                inner, meta = unwrap_subclass(out)
+                unwrapped_outputs.extend(inner)
+                output_subclass_metas.append((len(inner), meta))
+            else:
+                unwrapped_outputs.append(out)
+                output_subclass_metas.append((1, None))
+
+        return unwrapped_outputs
+
+    with fake_mode, preserve_node_meta():
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py:239]: comment: preserve_node_meta is needed for fx.traceback.annotate to work.

File: torchtitan/experiments/graph_trainer/nonstrict_tracer.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        output_tensors = outputs[idx : idx + num_tensors]
+        idx += num_tensors
+        if meta is None:
+            wrapped_outputs.append(output_tensors[0])
+        else:
+            wrapped_outputs.append(wrap_to_subclass(list(output_tensors), meta))
+    return wrapped_outputs
+
+
+def _remove_cpu_shadow_chains(gm: torch.fx.GraphModule) -> None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/nonstrict_tracer.py]: Add docstring for this .

DTensor shard propagation introduce some cpu ops during tracing, this function is a post_processing utils that remove it. 
This is kinda adhoc, We wonder if we have better to for tracer to avoid these in the first place?

File: torchtitan/experiments/graph_trainer/nonstrict_tracer.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    traced._params_spec = params_spec
+    traced._input_subclass_metas = subclass_metas
+    traced._output_subclass_metas = output_subclass_metas
+
+    return traced
+
+
+def run_traced_module(
+    traced: torch.fx.GraphModule,
+    mod: nn.Module,
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/nonstrict_tracer.py]: I don't think we should pass mod into this function, we should pass params_and_buffers in another way. Export had some old wisdom for how to handle this.

File: torchtitan/experiments/graph_trainer/nonstrict_tracer.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    traced._params_len = params_len
+    traced._params_spec = params_spec
+    traced._input_subclass_metas = subclass_metas
+    traced._output_subclass_metas = output_subclass_metas
+
+    return traced
+
+
+def run_traced_module(
+    traced: torch.fx.GraphModule,
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/nonstrict_tracer.py]: renamed to traced_gm.

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    def __init__(self, model, loss_fn):
+        super().__init__()
+        self.model = model
+        self.loss_fn = loss_fn
+
+    def forward(self, *args):
+        *fwd_args, labels = args
+        logits = self.model(*fwd_args)
+        loss = self.loss_fn(logits, labels)
+        params = [p for _, p in self.model.named_parameters(remove_duplicate=False)]
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py:38]: move to init

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            0, 256, (self.BATCH_SIZE, self.SEQ_LEN), device=self.DEVICE
+        )
+        return model, tokens, labels
+
+    def test_mlp_forward(self):
+        model, tokens, labels = self._make_mlp()
+        traced = trace_module(model, (tokens,))
+        out_eager = model(tokens)
+        out_traced = run_traced_module(traced, model, (tokens,))
+        out_traced = rewrap_outputs(out_traced, traced._output_subclass_metas)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py]: rewrap_outputs should be an internal step of run_traced_module

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    def _make_mlp(self):
+        model = SimpleMLP().to(device=self.DEVICE, dtype=self.DTYPE)
+        tokens = torch.randint(
+            0, 256, (self.BATCH_SIZE, self.SEQ_LEN), device=self.DEVICE
+        )
+        labels = torch.randint(
+            0, 256, (self.BATCH_SIZE, self.SEQ_LEN), device=self.DEVICE
+        )
+        return model, tokens, labels
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py]: move get_loss into this _make_mlp, and return loss_fn

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    BATCH_SIZE = 2
+    SEQ_LEN = 128
+    NUM_STEPS = 5
+    LR = 1e-3
+
+    def setUp(self):
+        # Required by torch.use_deterministic_algorithms(True) — cuBLAS needs
+        # a larger workspace to select deterministic kernels.
+        self._prev_cublas = os.environ.get("CUBLAS_WORKSPACE_CONFIG")
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py]: really, do we need these?

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            grads_tr = wrapped[1:]
+            for p, g in zip(model_copy.parameters(), grads_tr):
+                p.grad = g
+            opt_copy.step()
+            opt_copy.zero_grad()
+
+            self.assertTrue(
+                torch.equal(loss_ref, loss_tr),
+                f"Step {step}: loss mismatch",
+            )
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py]: Also check gradient bitwise.

File: torchtitan/experiments/graph_trainer/nonstrict_tracer.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    full_args = tuple(params_and_buffers_flat) + tuple(user_args_flat)
+
+    unwrapped_args = []
+    subclass_metas = []
+
+    for arg in full_args:
+        if isinstance(arg, torch.Tensor) and is_traceable_wrapper_subclass(arg):
+            inner_tensors, meta = unwrap_subclass(arg)
+            unwrapped_args.extend(inner_tensors)
+            subclass_metas.append((len(inner_tensors), meta))
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/nonstrict_tracer.py]: Give subclass_metas a structure

File: torchtitan/experiments/graph_trainer/nonstrict_tracer.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    with fake_mode, preserve_node_meta():
+        traced = make_fx(fn_with_subclass_handling, record_stack_traces=True)(
+            *fake_args
+        )
+
+    _remove_cpu_shadow_chains(traced)
+
+    traced._params_len = params_len
+    traced._params_spec = params_spec
+    traced._input_subclass_metas = subclass_metas
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/nonstrict_tracer.py]: these 3 are not used, are they needed?

**Comment by SherlockNoMad** [comment]
@claude

This PR is AI generated, review carefully. 
- Look for opportunities to refactor and simplify.
- Does the tracer has a clear API and user contract? 
- Need docstring for every user facing API. 
- The input/output subclass wrapping logic should be more structured.

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py (hunk 8 above)
Comment (yiming0416) [reply torchtitan/experiments/graph_trainer/tests/test_trace_module.py]: Can we directly use `set_determinism` from `torchtitan/distributed/utils.py`?

**Comment by SherlockNoMad** [comment]
Also plz include test with Dtensor weight and inputs.

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py (hunk 5 above)
Comment (ydwu4) [reply torchtitan/experiments/graph_trainer/tests/test_trace_module.py:38]: We can't store self.params in __init__ because we need to look up the fake parameters and buffers that are swaped and use the accociated proxy for tracing. Stord self.params will become tensor constants because they refer the the parameters before to_fake and is not tracked by proxy. Added a comment!

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+def _remove_cpu_shadow_chains(gm: torch.fx.GraphModule) -> None:
+    """Remove dead CPU tensor chains left by DTensor's shadow-op bookkeeping.
+
+    DTensor keeps CPU "shadow" copies of tensor metadata (size, stride) as
+    regular aten ops.  After make_fx tracing these ops end up in the graph but
+    never feed a real GPU computation, so they are pure overhead.  This pass
+    finds every chain rooted at a CPU ``empty_strided`` whose outputs never
+    reach a GPU node with downstream users, and erases the whole chain.
+
+    TODO: upstream this into a general FX dead-code-elimination pass.
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: I tried general FX dead-code-elimination pass, and it doesn't work for inplace op as it's impure. 

But we actually remove cpu shadow node, even if it's inplce op.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+    """Holds the traced graph and metadata needed to run it."""
+
+    gm: torch.fx.GraphModule
+    params_len: int
+    params_spec: pytree.TreeSpec
+    input_subclass_layouts: list[SubclassLayout]
+    output_subclass_layouts: list[SubclassLayout]
+
+
+def unwrap_subclass(t: torch.Tensor) -> tuple[list[torch.Tensor], SubclassMeta | None]:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: _unwrap_subclass, I suppose this not use facing.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+        attrs=attrs,
+        ctx=ctx,
+        inner_metas=inner_metas,
+        outer_size=t.size(),
+        outer_stride=t.stride(),
+    )
+    return all_inner, meta
+
+
+def wrap_to_subclass(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: _wrap_to_subclass

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py (hunk 8 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/tests/test_trace_module.py]: yes
